### PR TITLE
[boost] Update to v1.84.0

### LINKED ISSUE #34982: [boost-beast] update to 1.84
### Library name

boost-beast

### New version number

1.84

### Other information that may be useful (release notes, etc...)

Building an application that uses boost-beast with Xcode 15 generates a compiler error:
error: call to 'async_teardown' is ambiguous

This problem is already fixed with https://github.com/boostorg/beast/pull/2683.

The current version of boost is 1.83.  The vcpkg port needs to update to 1.84 as soon as it is available.

### LINKED ISSUE #35187: boost-pfr dependency isn't sufficient to use Boost::pfr
**Describe the bug**
Add "boost-pfr" to vcpkg.json dependencies, can't use the library.

**Environment**
- OS: ubuntu 23.04
- Compiler: gcc 13.2
- OS: Windows 11
- Compiler: Microsoft (R) C/C++ Optimizing Compiler Version 19.37.32825 for x64

(note boost-pfr is header-only, so no compilation involved)

**To Reproduce**
1. Add "boost-pfr" to vcpkg.json dependencies for a project
2. Configure
3. `find_package(Boost)` fails with `Could NOT find Boost (missing: Boost_INCLUDE_DIR)`
4. Manually add "boost-config" to vcpkg.json dependencies
5.  `find_package(Boost)` now succeeds
6.  but `Boost::pfr` target not found (`Target links to: Boost::pfr but the target was not found`)

**Expected behavior**
Specify "boost-pfr" alone in vcpkg.json dependencies for a project,  `find_package(Boost)` succeeds and `Boost::pfr` target available for us in `target_link_libraries`

**Additional context**
Lack of `Boost::pfr` target can be worked around with

```
    if(NOT TARGET Boost::pfr)
        add_library(Boost::pfr INTERFACE IMPORTED)
        set_target_properties(Boost::pfr PROPERTIES INTERFACE_INCLUDE_DIRECTORIES ${Boost_INCLUDE_DIRS})
    endif()
```

Note also that a Fetch Content of the git repo https://github.com/boostorg/pfr.git works perfectly, providing the target `Boost::pfr` out-the-box, and this is in fact the easiest way to get Boost pfr at the moment.

It would be great for this to be fixed for the forthcoming Boost 1.84 release, as Boost::pfr includes the ability to automatically get field names; there is a lot of community interest.

### LINKED ISSUE #35685: [boost] update to 1.84.0
### Library name

boost

### New version number

1.84.0

### Other information that may be useful (release notes, etc...)

https://github.com/boostorg/boost/releases/tag/boost-1.84.0

https://www.boost.org/users/history/version_1_84_0.html

### PR DESCRIPTION
**Pull request by yurybura**
Release Notes: https://www.boost.org/users/history/version_1_84_0.html

Fixes #35685
Fixes #35187
Fixes #34982

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/boost-accumulators/portfile.cmake [hunk 1]
```diff
@@ -1,10 +1,11 @@
 # Automatically generated by scripts/boost/generate-ports.ps1
+vcpkg_minimum_required(VERSION 2022-10-12)
```
Comment (Neumann-A) [inline ports/boost-accumulators/portfile.cmake:2]: Why is this here? This is already required in https://github.com/microsoft/vcpkg/blob/eb33d2f7583405fca184bcdf7fdd5828ec88ac05/scripts/ports.cmake#L110C1-L110C1

Comment (yurybura) [reply ports/boost-accumulators/portfile.cmake:2]: I didn't know that this line is redundant. Will be removed.

Comment (yurybura) [reply ports/boost-accumulators/portfile.cmake]: Fixed.

**Comment by BillyONeal** [comment]
I kicked the macos machine that died but not restarting build since there were other failures

File: ports/gtsam/portfile.cmake [hunk 2]
```diff
@@ -21,6 +21,7 @@ vcpkg_cmake_configure(
         -DGTSAM_USE_SYSTEM_METIS=On
         -DGTSAM_INSTALL_CPPUNITLITE=OFF
         -DGTSAM_BUILD_TYPE_POSTFIXES=OFF
+        -DCMAKE_CXX_STANDARD=11 # Boost v1.84.0 libraries require C++11
```
Comment (Neumann-A) [inline ports/gtsam/portfile.cmake:24]: wrong approach. Instead make the wrapper set the minimum standard.

Comment (yurybura) [reply ports/gtsam/portfile.cmake:24]: Sorry, I don't understand what you mean.  Could you suggest your approach?

Comment (Neumann-A) [reply ports/gtsam/portfile.cmake:24]: You are updating N ports because the usage requirement of boost changed. It is better to change FindBoost.cmake to carry the usage requirement instead of patching into every corner of the universe. So vcpkg-cmake-wrapper-cmake needs to be used for Boost and its targets (e.g. via target_compile_features(BOOST_TARGET INTERFACE cxx11 [or however the feature is named in cmake]) ).

Comment (yurybura) [reply ports/gtsam/portfile.cmake:24]: I thought about this. 1) FindBoost.cmake is part of standard CMake modules. 2) I didn't find goodness way to make interface requirements for every Boost library.
3) Several ports fail and only with AppleClang. I decided to patch these ports.

Please suggest your code if you know a better solution.

Comment (Neumann-A) [reply ports/gtsam/portfile.cmake:24]: > 2) I didn't find goodness way to make interface requirements for every Boost library.

There is a `vcpkg-cmake-wrapper.cmake` for boost in `boost-uninstall` if you add the requirement to the headers target it will probably stick to all other boost components. Secondly you can inspect `CMAKE_CXX_STANDARD` in the wrapper and raise it if necessary.

Comment (yurybura) [reply ports/gtsam/portfile.cmake:24]: I don't think that it is a good solution. A good solution will be like that `target_compile_features(bar PUBLIC cxx_std_11)`, but targets handled by standard cmake finder.
Finally, every library should be fixed to compile with the new Boost without vcpkg.

File: ports/kenlm/portfile.cmake [hunk 3]
```diff
@@ -25,6 +25,7 @@ vcpkg_cmake_configure(
         -DFORCE_STATIC=OFF #already handled by vcpkg
         -DENABLE_PYTHON=OFF # kenlm.lib(bhiksha.cc.obj) : fatal error LNK1000: Internal error during IMAGE::Pass2
         -DCOMPILE_TESTS=OFF
+        -DCMAKE_CXX_STANDARD=11 # Boost v1.84.0 libraries require C++11
```
Comment (dg0yt) [inline ports/kenlm/portfile.cmake:28]: Also in #35907 - but because it doesn't build with C++17.

File: ports/boost-cobalt/vcpkg.json [hunk 4]
```diff
@@ -5,6 +5,7 @@
   "description": "Boost cobalt module",
   "homepage": "https://github.com/boostorg/cobalt",
   "license": "BSL-1.0",
+  "supports": "!ios & !android",
```
Comment (dg0yt) [inline ports/boost-cobalt/vcpkg.json:8]: Is this a general incompatibility, or just a vcpkg CI / NDK r25 problem?

Comment (yurybura) [reply ports/boost-cobalt/vcpkg.json:8]: > Is this a general incompatibility, or just a vcpkg CI / NDK r25 problem?

This is not a general incompatibility. In both cases (iOS and Android) a Clang 13 based compiler is used, which is supported by Boost.Cobalt. The main problem is that the included libcxx++ does not contain the Concepts library.

File: ports/kenlm/portfile.cmake (hunk 3 above)
Comment (yurybura) [reply ports/kenlm/portfile.cmake:28]: This is workaround to fix CI. 

I think vcpkg should have a policy to stop supporting (updating) legacy libraries because they block other ports from being updated. Updating Boost in this case is quite painful..

Comment (yurybura) [reply ports/kenlm/portfile.cmake]: The kenlm changes have been reverted and synchronized with master.

**Comment by yurybura** [comment]
@BillyONeal CI build is hung.

**Review by Adela0814 — DISMISSED** [review]

**Comment by BillyONeal** [comment]
I was forced to cancel this build as it's the last thing holding the before-patch-tuesday machines alive. I tried to merge with master for you but permission is denied for me to do so.

**Comment by yurybura** [comment]
@BillyONeal Permission has been granted. Sorry I forgot about this check box..

**Comment by BillyONeal** [comment]
> @BillyONeal Permission has been granted. Sorry I forgot about this check box..

No problem!

**Comment by yurybura** [comment]
@BillyONeal The `master` branch is suddenly broken. Each update of this PR requires a long rebuild of all checks (up to 16 hours). My changes have already passed all checks before. Could you start the review? Thanks!)

**Comment by dg0yt** [comment]
> @BillyONeal The `master` branch is suddenly broken. Each update of this PR requires a long rebuild of all checks (up to 16 hours). My changes have already passed all checks before. Could you start the review? Thanks!)

I don't think it is broken. But you may want to factor out the `vcpkg-cmake-get-vars` change. If it is really necessary to have it.

**Comment by dg0yt** [comment]
And there is https://github.com/microsoft/vcpkg/pull/36259.

**Comment by yurybura** [comment]
> I don't think it is broken. But you may want to factor out the `vcpkg-cmake-get-vars` change. If it is really necessary to have it.

It is used to detect compiler version in [ports/boost-cobalt/b2-options.cmake](https://github.com/microsoft/vcpkg/pull/35693/files#diff-8b01d53debb60fc04e1d83dc27e35e0568b966239f22463f6c755568b1bc3f73).

> And there is #36259.

Yes I saw it. The local build of `netgen` without my changes also failed, but it breaks the CI checks of this PR.

**Comment by BillyONeal** [comment]
> @BillyONeal The `master` branch is suddenly broken. Each update of this PR requires a long rebuild of all checks (up to 16 hours). My changes have already passed all checks before. Could you start the review? Thanks!)

Under some circumstances we might be able to do special dispensation to ignore CI failures, but given that we had a compiler version change out from under this I'm not really comfortable skipping PR build.

**Comment by yurybura** [comment]
> > @BillyONeal The `master` branch is suddenly broken. Each update of this PR requires a long rebuild of all checks (up to 16 hours). My changes have already passed all checks before. Could you start the review? Thanks!)
> 
> Under some circumstances we might be able to do special dispensation to ignore CI failures, but given that we had a compiler version change out from under this I'm not really comfortable skipping PR build.

Sure, thanks!

**Comment by yihuajack** [comment]
Boost version on https://vcpkg.io/en/packages is still not updated (1.83.0)
